flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -89,7 +90,7 @@
     private final KafkaRecordSerializationSchema<IN> recordSerializer;
     private final Callback deliveryCallback;
     private final KafkaRecordSerializationSchema.KafkaSinkContext kafkaSinkContext;
-
+    private final AtomicReference<Exception> asyncProducerException = new AtomicReference<>();
```

Comment:
The fix overall makes sense to me, although I'm wondering if we can avoid using an `AtomicReference` and just use a `volatile` reference instead? That would be better for overhead reasons because this reference lookup happens on the critical path on every `write()`. Using `AtomicReference` you essentially have this extra pointer chase to the wrapped `volatile` variable.

It looks like the only functionality that requires the solution to use an `AtomicReference` is the `compareAndSet` method in the callback, to make sure that we only ever update the reference once. I think relaxing this is ok. wdyt @mas-chen?

Reply:
Discussed with Gordon offline. We agreed to do this to optimize the write path. The assumption is that the change will still surface the first exception within a group of write requests. This is possible due to the exception handling logic in write/flush/WriterCallback.